### PR TITLE
Refactor for namespace support.

### LINKED ISSUE #13: Namespaces not supported?
Great work, I'm eager to use it!

I'm working on a large browser based project with namespaces. It looks like tplant outputs only the startuml and enduml tags when hitting the keyword namespace or module,..

### PR DESCRIPTION
**Pull request by bafolts**
This will add support for namespaces and close #13

Separate component classes will be created for `Namespace`, `File`, `Class`, `Enum`, `Interface`, `Method`, `Property`, etc. The first step will then walk the typescript abstract syntax tree and create the appropriate component classes to `File`. From there the second pass will convert each `File` to it's corresponding plantuml.

### CODE REVIEW

**Comment by bafolts** [comment]
@marcosvrs I haven't added support for namespaces yet, but this pull request has the large refactor in place. After this it should be much simpler to add namespace. I will finish the work for namespace tomorrow. If this gets merged I will do it in a separate branch to create a smaller pull request otherwise I will continue the work on this branch.

**Comment by marcosvrs** [comment]
I have some suggestions:
* Add method `toString()` (or `toPUML()`) to each component composites, so it can handle the conversion by their own. And convert their members without knowing their logic.
* Separate also all the functions inside the `generateDocumentation.ts` to component factories, for a better separation.
* Work with interfaces for inversion of control. In the scope of my suggestion: the factories would be statics and return an interface references of each object. Example:
```typescript
interface IComponentComposite {
    toPUML(): string;
}
class ClassComposite implements IComponentComposite {...}
interface IComponentFactory {
    serialize(node: ts.Node): IComponentComposite | undefined;
}
class ClassFactory implements IComponentFactory {...}
```
Later, you can easily extend the logic for the `NamespaceComposite` and `NamespaceFactory` without much work.
At the end, you can even combine the `generateDocumentation.ts` and the `convertToPlant.ts` with simple functions since they'll be smaller.

PS: we need to think already how to compile all TS files in just one JS file. Webpack? Gulp?

**Comment by bafolts** [comment]
Webpack works fine. I will implement these suggestions, once they are in place the webpack work and other improvements can be done in separate pull requests.

**Comment by bafolts** [comment]
@marcosvrs I have made the requested changes. This should be fine to merge now and then build upon the namespace support once merged.

File: src/components/ClassFactory.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
+
+import ts from 'typescript';
+import { Class } from './Class';
+import * as ComponentFactory from './ComponentFactory';
+
+export function create(classSymbol: ts.Symbol, checker: ts.TypeChecker): Class {
```
Comment (marcosvrs) [inline src/components/ClassFactory.ts]:
```suggestion
export function create(classSymbol: ts.Symbol, checker: ts.TypeChecker): IComponentComposite {
```

File: src/components/File.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+import { Interface } from './Interface';
+import { Namespace } from './Namespace';
+
+/**
+ * Represents the metadata for a file containing typescript
+ */
+export class File implements IComponentComposite {
+    public componentKind: ComponentKind = ComponentKind.FILE;
+    public name: string = '';
+    public parts: (Namespace | Class | Interface | Enum)[] = [];
```
Comment (marcosvrs) [inline src/components/File.ts]:
```suggestion
    public parts: IComponentComposite[] = [];
```

File: src/components/Namespace.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+import { Method } from './Method';
+import { Property } from './Property';
+
+/**
+ * Represents the metadata for a namespace within typescript
+ */
+export class Namespace implements IComponentComposite {
+    public name: string = '';
+    public componentKind: ComponentKind = ComponentKind.NAMESPACE;
+    public parts: (Enum | Class | Interface | Namespace | Property | Method)[] = [];
```
Comment (marcosvrs) [inline src/components/Namespace.ts]:
```suggestion
    public parts: IComponentComposite[] = [];
```

File: src/components/ComponentFactory.ts [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+           declaration.kind === ts.SyntaxKind.GetAccessor ||
+           declaration.kind === ts.SyntaxKind.SetAccessor ||
+           declaration.kind === ts.SyntaxKind.Parameter;
+}
+
+export function isTypeParameter(declaration: ts.NamedDeclaration): boolean {
+    return declaration.kind === ts.SyntaxKind.TypeParameter;
+}
+
+export function getMemberModifierType(memberDeclaration: ts.Declaration): 'public' | 'private' | 'protected' {
```
Comment (marcosvrs) [inline src/components/ComponentFactory.ts]: Create a type for the return value

File: src/components/EnumFactory.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+                    result.push(EnumValueFactory.create(memberSymbol, memberDeclaration, checker));
+                }
+            });
+        });
+    }
+
+    return result;
+}
+
+export function create(enumSymbol: ts.Symbol, checker: ts.TypeChecker): Enum {
```
Comment (marcosvrs) [inline src/components/EnumFactory.ts]:
```suggestion
export function create(enumSymbol: ts.Symbol, checker: ts.TypeChecker): IComponentComposite {
```

File: src/components/InterfaceFactory.ts [hunk 6]
```diff
@@ -0,0 +1,29 @@
+
+import ts from 'typescript';
+import * as ComponentFactory from './ComponentFactory';
+import { Interface } from './Interface';
+
+export function create(interfaceSymbol: ts.Symbol, checker: ts.TypeChecker): Interface {
```
Comment (marcosvrs) [inline src/components/InterfaceFactory.ts]:
```suggestion
export function create(interfaceSymbol: ts.Symbol, checker: ts.TypeChecker): IComponentComposite {
```

File: src/components/NamespaceFactory.ts [hunk 7]
```diff
@@ -0,0 +1,73 @@
+
+import ts from 'typescript';
+import * as ClassFactory from './ClassFactory';
+import * as ComponentFactory from './ComponentFactory';
+import * as EnumFactory from './EnumFactory';
+import * as InterfaceFactory from './InterfaceFactory';
+import { Namespace } from './Namespace';
+import * as NamespaceFactory from './NamespaceFactory';
+
+export function create(namespaceSymbol: ts.Symbol, checker: ts.TypeChecker): Namespace {
```
Comment (marcosvrs) [inline src/components/NamespaceFactory.ts]:
```suggestion
export function create(namespaceSymbol: ts.Symbol, checker: ts.TypeChecker): IComponentComposite {
```

**Pull request merged**